x/ecocredit/server/msg_server.go

```diff
@@ -577,3 +516,352 @@ func (s serverImpl) isCreatorAllowListed(ctx types.Context, allowlist []string,
+			}
+
+			// verify that seller's credit balance has quantity in the sell order
+			err = verifyCreditBalance(store, sellerAddr, sellOrder.BatchDenom, sellOrder.Quantity)
+			if err != nil {
+				return nil, ecocredit.ErrInvalidSellOrder.Wrap(err.Error())
+			}
+
+			// get decimal amount of credits available
+			creditsAvailable, err := math.NewDecFromString(sellOrder.Quantity)
+			if err != nil {
+				return nil, ecocredit.ErrInvalidSellOrder.Wrap(err.Error())
+			}
+
+			creditsToReceive := creditsDesired
+
+			// check if credits desired is more than credits available
+			if creditsDesired.Cmp(creditsAvailable) == -1 {
+
+				// error if partial fill disabled
+				if order.DisablePartialFill {
+					return nil, ecocredit.ErrInsufficientFunds.Wrap("sell order does not have sufficient credits to fill the buy order")
+				}
+
+				creditsToReceive = creditsAvailable
+
+				// recalculate coinsNeeded if creditsToReceive is not creditsDesired
+				coinToSend, err = getCoinNeeded(creditsToReceive, bidPrice)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// Move the coins to the seller account
+			err = s.bankKeeper.SendCoins(sdkCtx, buyerAddr, sellerAddr, sdk.Coins{coinToSend})
+			if err != nil {
+				return nil, err
+			}
+
+			credit := &ecocredit.MsgSend_SendCredits{
+				BatchDenom:     sellOrder.BatchDenom,
+				TradableAmount: creditsToReceive.String(),
+				// TODO: handle auto-retire settings #621
+				//RetiredAmount: creditsToRetire.String(),
+				//RetirementLocation: retirementLocation,
+			}
+
+			// Move the credits to the buyer account
+			err = s.sendEcocredits(ctx, credit, store, sellerAddr, buyerAddr)
+			if err != nil {
+				return nil, err
+			}
+
+			buyOrderID := s.buyOrderSeq.NextVal(ctx)
+			buyOrderIds = append(buyOrderIds, buyOrderID)
+
+			err = ctx.EventManager().EmitTypedEvent(&ecocredit.EventBuy{
+				BuyOrderId:         buyOrderID,
+				SellOrderId:        sellOrderId,
+				Quantity:           order.Quantity,
+				BidPrice:           order.BidPrice,
+				DisableAutoRetire:  order.DisableAutoRetire,
+				DisablePartialFill: order.DisablePartialFill,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+		// TODO: implement processing for filter option #623
+		//case *ecocredit.MsgBuy_Order_Selection_Filter:
+
+		default:
+			return nil, sdkerrors.ErrInvalidRequest
+		}
+
+		ctx.GasMeter().ConsumeGas(gasCostPerIteration, "create buy order")
+	}
+
+	return &ecocredit.MsgBuyResponse{BuyOrderIds: buyOrderIds}, nil
+}
+
+// AllowAskDenom adds a new ask denom
+func (s serverImpl) AllowAskDenom(goCtx context.Context, req *ecocredit.MsgAllowAskDenom) (*ecocredit.MsgAllowAskDenomResponse, error) {
+	ctx := types.UnwrapSDKContext(goCtx)
+
+	if req.RootAddress != authtypes.NewModuleAddress(govtypes.ModuleName).String() {
+		return nil, sdkerrors.ErrUnauthorized.Wrap("root address must be governance module address")
+	}
```

Comment:
@aaronc @blushi Can one of you confirm that this is how the root address stuff is supposed to work?